### PR TITLE
Add UEA multivariate time series dataset wrapper

### PR DESCRIPTION
**Pull request by Arsharathi1**
## Description

This PR introduces support for UEA Multivariate Time Series (MTS) datasets in the dataset pipeline.

The implementation:

* Adds a wrapper for loading UEA MTS datasets.
* Parses the UEA `.ts` file format.
* Supports training, validation, and test splits.
* Handles multivariate time series with different sequence lengths.
* Maps dataset labels to numerical class indices.
* Integrates the datasets with the existing parsing workflow.

The implementation has been tested on UEA multivariate time series datasets to verify dataset loading, parsing, label handling, and dataset preparation.

### CODE REVIEW

File: peepholelib/datasets/UEAdataset.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        return len(self.X)
+    def __getitem__(self, idx):
+        x = self.X[idx]
+        if self.transform is not None:
+            x = self.transform(x)
+        return {
+            "timeseries": x,
+            "label": self.y[idx]
+        }
+# UEA DATA WRAPPER
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: put the comment on the line immediately before the class.

File: peepholelib/datasets/UEAdataset.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        )
+        self.train_ratio = kwargs.get(
+            "train_ratio",
+            0.8
+        )
+        self.seed = kwargs.get(
+            "seed",
+            42
+        )
+# Variable-length handling
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Indent your comments to the same level as the code they refer to.

File: peepholelib/datasets/UEAdataset.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        self.variable_length = kwargs.get(
+            "variable_length",
+            "pad"
+        )
+
+        valid_modes = {
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Why is this a set?, It could be a simple list.

File: peepholelib/datasets/UEAdataset.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+        if path is None:
+            raise ValueError(
+                "UEA dataset path must be provided."
+            )
+
+        self.root = Path(path)
+        self.dataset_name = self.root.name
+
+        # transforms
+        self.transform = kwargs.get(
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Please make this `kwargs.get()` to be a single line of code

File: peepholelib/datasets/UEAdataset.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+
+        if self.variable_length not in valid_modes:
+            raise ValueError(
+                f"Unknown variable_length mode: "
+                f"{self.variable_length}"
+            )
+
+        if (
+            self.transform is not None
+            and
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Should not be checking if it is callable here. This will be done in the transformation's `Compose()`

File: peepholelib/datasets/UEAdataset.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+            self.augmentation is not None
+            and
+            not callable(self.augmentation)
+        ):
+            raise TypeError(
+                "augmentation must be callable."
+            )
+
+        self.__dataset__ = {}
+        self.__load_data__()
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: why are you calling the `__load_data__()` here? the parsing functions should call it

File: peepholelib/datasets/UEAdataset.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: add a return call

File: peepholelib/datasets/UEAdataset.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+            )
+
+        if (
+            self.augmentation is not None
+            and
+            not callable(self.augmentation)
+        ):
+            raise TypeError(
+                "augmentation must be callable."
+            )
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Where are the compositions of the transformations and augmentations? if they are not used at all, you can delete them.

File: peepholelib/datasets/UEAdataset.py [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+
+        self.__dataset__ = {}
+        self.__load_data__()
+        
+# LOAD DATA
+    def __load_data__(self):
+
+        print("\n==============================")
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: Remove the debugging prints.

File: peepholelib/datasets/UEAdataset.py [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+        print("LOADING DATASET:", self.dataset_name)
+        print("PATH:", self.root)
+
+        train_file = (
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: make this to become one line and remove the unnecessary parenthesis

File: peepholelib/datasets/UEAdataset.py [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+            self.root /
+            f"{self.dataset_name}_TRAIN.ts"
+        )
+
+        test_file = (
+            self.root /
+            f"{self.dataset_name}_TEST.ts"
+        )
+
+        if not train_file.exists():
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: The first thing that `load_ts()` does is to read the file. If the file does not exist there, you will get the same error here.

If checking for an error does not save you anything, there is no reason to make the check. Please remove these file checks.

File: peepholelib/datasets/UEAdataset.py [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+            raise FileNotFoundError(
+                f"TRAIN file not found:\n{train_file}"
+            )
+
+        if not test_file.exists():
+            raise FileNotFoundError(
+                f"TEST file not found:\n{test_file}"
+            )
+# LOAD RAW TRAIN
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: indent all comments

File: peepholelib/datasets/UEAdataset.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+        print("NOW LOADING TRAIN:", train_file)
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: remove all checking prints

File: peepholelib/datasets/UEAdataset.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+        train_samples, train_y = self._load_ts(
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: if functions do not have may parametes, call them in a single line.

File: peepholelib/datasets/UEAdataset.py [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+            target_length
+        )
+# CREATE DATASETS
+        full_train_ds = TSDataset(
+            train_X,
+            train_y,
+            transform=self.augmentation
+        )
+
+        n_total = len(full_train_ds)
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: if you are training the models, you will need a validation split as well, so define the ratios as an array of size 3, check in the instantiation if they sum to 1, and make a single for loop for creating the splits, instead of repeating the code 3 times.

File: peepholelib/datasets/UEAdataset.py [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+# Parse one sample
+                parts = line.split(":")
+                if len(parts) < 2:
+                    continue
+                raw_y.append(
+                    parts[-1].strip()
+                )
+                sample = []
+                for dim in parts[:-1]:
+                    values = np.asarray(
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: should not use numpy, use alweays torch. make this into one line

File: peepholelib/datasets/UEAdataset.py [hunk 17]
```diff
@@ -0,0 +1,441 @@
...
+        print("Validation samples:", len(val_ds))
+        print("Test samples:", len(test_ds))
+        print("==============================\n")
+
+# TS FILE PARSER
+    def _load_ts(self, file_path):
+        raw_X = []
+        raw_y = []
+        reading_data = False
+        with open(file_path, "r") as f:
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: You should be reading lines until you find the data; after that, you should read all the lines in one go, and process them all together instead of reading line by line. Note that some functions read multiple lines of the text. So you won't need all those nested loops.

File: peepholelib/datasets/UEAdataset.py [hunk 18]
```diff
@@ -0,0 +1,441 @@
...
+                    f"{self.dataset_name}"
+                )
+        y = np.asarray(
+            [
+                label_map[label]
+                for label in raw_y
+            ],
+            dtype=np.int64
+        )       
+# DEBUG
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: remove the debug stuff for submission

File: peepholelib/datasets/UEAdataset.py [hunk 19]
```diff
@@ -0,0 +1,441 @@
...
+                        dtype=np.float32
+                    )
+                    sample.append(values)
+                raw_X.append(sample)
+        if len(raw_X) == 0:
+            raise RuntimeError(
+                f"No samples parsed from {file_path}"
+            )
+# LABEL ENCODING
+        if "TRAIN" in file_path.name.upper():
```
Comment (lsrosa) [inline peepholelib/datasets/UEAdataset.py]: instead of having a class with class methods, make a simple class and instantiate it in the init function. This is overcomplicating things, and it stinks of LLM garbage.

TBH, you should just define the registry as a dictionary in the __init__ and roll with it, no need for an extra class.

**Review by lsrosa — CHANGES_REQUESTED** [review]
Too many problems, please fix these first, then we can check the logic.

File: peepholelib/datasets/UEAdataset.py (hunk 3 above)
Comment (Arsharathi1) [reply peepholelib/datasets/UEAdataset.py]: Yes, I can use list here. As I am just checking unique item exist or not therefore I used set.

File: peepholelib/datasets/UEAdataset.py (hunk 15 above)
Comment (Arsharathi1) [reply peepholelib/datasets/UEAdataset.py]: For each sub datasets there is train.ts and test.ts files are given. In above script I have left test data as untouched and split the train data as 80 % of training data and 20 % of validation data. How to apply the ratios  as an array of size 3 to data as it is already given in separate train.ts and test.ts files?

Comment (lsrosa) [reply peepholelib/datasets/UEAdataset.py]: As in the other datasets, a separate part of the training set is used for validation. The split will not be 3 values; you are right there, but in any case, you should create 3 splits.
